File: supabase/functions/_shared/chat_map_link.ts
import type { ListingDetail } from "./chat_answer_openai.ts";
import type { BotReply, ChatLink } from "./chat_logic.ts";

export type MapPin = {
  lat: number;
  lng: number;
  label: string;
  btsStation?: string | null;
  maxDistanceBtsKm?: number | null;
};

const LOCATION_KEYS = [
  "โลเคชั่น",
  "location",
  "แผนที่",
  "แมพ",
  "map",
  "google map",
  "ใกล้ bts",
  "ใกล้ mrt",
  "ทำเล",
  "เดินทาง",
  "รถไฟฟ้า",
  "กี่เมตร",
  "ห่าง bts",
  "ห่าง mrt",
  "ขอพิกัด",
  "พิกัด",
  "ใกล้สถานี",
  "เดินถึง",
  "ใกล้ห้าง",
];

export function isLocationQuestion(text: string): boolean {
  const q = text.toLowerCase();
  return LOCATION_KEYS.some((k) => q.includes(k));
}

export function buildGoogleMapsUrl(pin: { lat: number; lng: number }): string {
  return `https://www.google.com/maps/search/?api=1&query=${pin.lat},${pin.lng}`;
}

export function locationMapLink(pin: MapPin): ChatLink {
  return {
    label: "เปิด Google Maps · โครงการ",
    kind: "viewing_location",
    listingId: "",
    refCode: buildGoogleMapsUrl(pin),
    projectName: pin.label,
  };
}

export function listingToMapPin(
  listing?: ListingDetail | null,
): MapPin | null {
  const lat = listing?.map_lat ?? null;
  const lng = listing?.map_lng ?? null;
  if (lat == null || lng == null || !Number.isFinite(lat) || !Number.isFinite(lng)) {
    return null;
  }
  return {
    lat,
    lng,
    label: listing?.project_name?.trim() || listing?.title?.trim() || "โครงการ",
    btsStation: listing?.project_bts ?? null,
    maxDistanceBtsKm: listing?.max_distance_bts_km ?? null,
  };
}

export function locationReplyTexts(pin: MapPin | null): string[] {
  if (!pin) {
    return [
      "แผนที่ที่แสดงจะบอกโซนโดยประมาณค่ะ หากต้องการทราบระยะทางจาก BTS/MRT แจ้งได้เลยนะคะ",
    ];
  }
  return [
    "แอดมินส่งโลเคชั่นโครงการให้ทางนี้นะคะ ลูกค้าสามารถคลิกเพื่อตรวจสอบพิกัดได้เลยค่ะ",
  ];
}

export function locationReplyLinks(pin: MapPin | null): ChatLink[] {
  if (!pin) return [];
  return [locationMapLink(pin)];
}

export function locationFaqReply(pin: MapPin | null): BotReply {
  if (!pin) {
    return {
      role: "ai",
      text:
        "แผนที่ที่แสดงจะบอกโซนโดยประมาณค่ะ หากต้องการทราบระยะทางจาก BTS/MRT " +
        "แจ้งได้เลยนะคะ แอดมินจะตรวจสอบรายละเอียดให้ค่ะ",
    };
  }

  const parts: string[] = [];
  if (pin.btsStation && pin.maxDistanceBtsKm != null && pin.maxDistanceBtsKm > 0) {
    const meters = Math.round(pin.maxDistanceBtsKm * 1000);
    parts.push(`ระยะทางจาก ${pin.btsStation} ในระบบแจ้งไว้ประมาณ ${meters} เมตรค่ะ`);
  } else if (pin.btsStation) {
    parts.push(`โครงการอยู่ใกล้ ${pin.btsStation} ค่ะ`);
  }
  parts.push(
    "แอดมินส่งโลเคชั่นโครงการให้ทางนี้นะคะ ลูกค้าสามารถคลิกเพื่อตรวจสอบพิกัดได้เลยค่ะ",
  );

  return {
    role: "ai",
    text: parts.join(" "),
    links: [locationMapLink(pin)],
  };
}

File: mobile/lib/utils/chat_map_link.dart
import '../models/chat_message.dart';
import '../l10n/app_strings.dart';

typedef ListingMapCoords = ({double? lat, double? lng});

const _locationKeys = [
  'โลเคชั่น',
  'location',
  'แผนที่',
  'แมพ',
  'map',
  'google map',
  'ใกล้ bts',
  'ใกล้ mrt',
  'ทำเล',
  'เดินทาง',
  'รถไฟฟ้า',
  'กี่เมตร',
  'ห่าง bts',
  'ห่าง mrt',
  'ขอพิกัด',
  'พิกัด',
  'ใกล้สถานี',
];

bool isLocationQuestion(String text) {
  final q = text.toLowerCase();
  return _locationKeys.any(q.contains);
}

const _internetKeys = [
  'อินเทอร์เน็ต',
  'เน็ต',
  'wifi',
  'wi-fi',
  'internet',
  'ติดเน็ต',
  'ติดตั้งเน็ต',
];

bool isInternetQuestion(String text) {
  final q = text.toLowerCase();
  return _internetKeys.any(q.contains);
}

const internetFaqBurst = [
  'ตามรายละเอียดค่าเช่าจะไม่ได้รวมอินเทอร์เน็ตค่ะ ทางเรามีบริการช่างอินเทอร์เน็ตให้นะคะ ทั้งสองค่าย',
  'ลูกค้าแค่ทำการเลือกแพ็กเกจ ไม่ต้องลำบากติดต่อพนักงานเองเลยค่ะ',
];

String? googleMapsUrlForCoords(double? lat, double? lng) {
  if (lat == null || lng == null) return null;
  return 'https://www.google.com/maps/search/?api=1&query=$lat,$lng';
}

List<String> locationReplyTexts(ListingMapCoords? listing) {
  if (listing?.lat == null || listing?.lng == null) {
    return const [
      'แผนที่ที่แสดงจะบอกโซนโดยประมาณค่ะ หากต้องการทราบระยะทางจาก BTS/MRT แจ้งได้เลยนะคะ',
    ];
  }
  return const [
    'แอดมินส่งโลเคชั่นโครงการให้ทางนี้นะคะ ลูกค้าสามารถคลิกเพื่อตรวจสอบพิกัดได้เลยค่ะ',
  ];
}

List<ChatMessageLink> locationReplyLinks(ListingMapCoords? listing, AppStrings copy) {
  final url = listing == null
      ? null
      : googleMapsUrlForCoords(listing.lat, listing.lng);
  if (url == null) return const [];
  return [
    ChatMessageLink.viewingLocation(
      label: copy.chatLinkViewingLocation,
      mapsUrl: url,
    ),
  ];
}
